File: frontend/src/utils/distance.ts
const EARTH_RADIUS_KM = 6371;

export function haversineKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const dlat = toRad(lat2 - lat1);
  const dlon = toRad(lon2 - lon1);
  const a =
    Math.sin(dlat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dlon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}

export function estimateEtaMin(distanceKm: number, speedKph: number | null): number {
  const speed = speedKph && speedKph > 1 ? speedKph : 5;
  return Math.max(1, Math.round((distanceKm / speed) * 60));
}

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

File: backend/app/utils/distance.py
import math

EARTH_RADIUS_KM = 6371.0


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return EARTH_RADIUS_KM * c


def bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    dlon = math.radians(lon2 - lon1)
    y = math.sin(dlon) * math.cos(math.radians(lat2))
    x = math.cos(math.radians(lat1)) * math.sin(math.radians(lat2)) - math.sin(math.radians(lat1)) * math.cos(
        math.radians(lat2)
    ) * math.cos(dlon)
    return (math.degrees(math.atan2(y, x)) + 360) % 360


def estimate_eta_min(distance_km: float, speed_kph: float | None) -> int:
    if speed_kph and speed_kph > 1:
        speed = speed_kph
    else:
        speed = 5.0
    hours = distance_km / speed
    minutes = int(hours * 60)
    return max(1, minutes)
